File: 백준/Gold/11779. 최소비용 구하기 2/최소비용 구하기 2.py
import heapq

def dijkstra(start):
    heap = []
    heapq.heappush(heap, (0, start))
    result[start] = 0

    while heap:
        sk, k = heapq.heappop(heap)

        if sk>result[k]:
            continue

        for i in arr[k]:
            cost = sk+i[1]
            if cost<result[i[0]]:
                result[i[0]] = cost
                visited[i[0]] = k
                heapq.heappush(heap, (cost, i[0]))


n = int(input())
m = int(input())
arr = [[] for _ in range(n+1)]
for _ in range(m):
    a, b, c = map(int, input().split())
    arr[a].append((b, c))
start, fin = map(int, input().split())

inf = int(21e10)
result = [inf] * (n+1)
visited = [0] * (n+1)
dijkstra(start)

ans = []
tmp = fin
while tmp != start:
    ans.append(str(tmp))
    tmp = visited[tmp]

ans.append(str(start))
ans.reverse()

print(result[fin])
print(len(ans))
print(" ".join(ans))

File: 백준/Gold/11779. 최소비용 구하기 2/최소비용 구하기 2.java
import java.util.*;
import java.io.*;

public class Main {

    static int n;
    static int INF = 987654321;
    static int[] dist;
    static int[] route;
    static class Pair implements Comparable<Pair> {
        int y, cost;

        public Pair (int y, int cost) {
            this.y = y;
            this.cost = cost;
        }

        @Override
        public int compareTo(Pair o) {
            return Integer.compare(this.cost, o.cost);
        }
    }
    static int start, end;
    static ArrayList<ArrayList<Pair>> arr = new ArrayList<>();
    static boolean[] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        int m = Integer.parseInt(br.readLine());
        dist = new int[n+1];
        route = new int[n+1];
        visited = new boolean[n+1];
        Arrays.fill(dist, INF);
        for (int i = 0; i<=n; i++) {
            arr.add(new ArrayList<>());
        }
        StringTokenizer st;
        for (int i = 0; i<m; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int cost = Integer.parseInt(st.nextToken());
            arr.get(a).add(new Pair(b, cost));
        }
        st = new StringTokenizer(br.readLine());
        start = Integer.parseInt(st.nextToken());
        end = Integer.parseInt(st.nextToken());

        dijkstra();

        System.out.println(dist[end]);

        ArrayList<Integer> temp = new ArrayList<>();
        int cur = end;
        while (cur != 0) {
            temp.add(cur);
            cur = route[cur];
        }
        System.out.println(temp.size());
        for (int i = temp.size()-1; i>=0; i--) {
            System.out.print(temp.get(i) + " ");
        }
    }

    public static void dijkstra() {
        PriorityQueue<Pair> q = new PriorityQueue<>();
        q.add(new Pair(start, 0));
        dist[start] = 0;
        route[start] = 0;

        while (!q.isEmpty()) {
            Pair curr = q.poll();
            if (!visited[curr.y]) visited[curr.y] = true;
            else continue;

            for (Pair p: arr.get(curr.y)) {
                int py = p.y;
                int pcost = p.cost;

                if (dist[py] > dist[curr.y] + pcost) {
                    dist[py] = dist[curr.y] + pcost;
                    q.add(new Pair(py, dist[py]));
                    route[py] = curr.y;
                }

            }
        }
    }
}